crates/runtime/src/instance/allocator/pooling/index_allocator.rs

```diff
@@ -0,0 +1,408 @@
+    ///
+    /// On allocation, we first try to find a slot with affinity for
+    /// the given module ID, if any. If not, we pick a random slot
+    /// ID. This random choice is unbiased across all free slots.
+    ReuseAffinity {
+        // Free-list of all slots. We use this to pick a victim when
+        // we don't have an appropriate slot with the preferred
+        // affinity.
+        free_list: Vec<usize>,
+        // Invariant: any module ID in this hashmap must have a
+        // non-empty list of free slots (otherwise we remove it).
+        per_module: HashMap<CompiledModuleId, Vec<usize>>,
+        // The state of any given slot. Records indices in the above
+        // list (empty) or two lists (with affinity), and these
+        // indices are kept up-to-date to allow fast removal.
+        slot_state: Vec<SlotState>,
+    },
+}
+
+#[derive(Clone, Debug)]
+pub(crate) enum SlotState {
+    Taken,
+    Empty {
+        /// Index in the global free list. Invariant:
+        /// free_list[slot_state[i].free_list_index] == i.
+        free_list_index: usize,
+    },
+    Affinity {
+        module: CompiledModuleId,
+        /// Index in the global free list. Invariant:
+        /// free_list[slot_state[i].free_list_index] == i.
+        free_list_index: usize,
+        /// Index in a per-module free list. Invariant:
+        /// per_module[slot_state[i].module][slot_state[i].per_module_index]
+        /// == i.
+        per_module_index: usize,
+    },
+}
+
+impl SlotState {
+    /// Get the index of this slot in the global free list.
+    fn free_list_index(&self) -> usize {
+        match self {
+            &Self::Empty { free_list_index }
+            | &Self::Affinity {
+                free_list_index, ..
+            } => free_list_index,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Update the index of this slot in the global free list.
+    fn update_free_list_index(&mut self, index: usize) {
+        match self {
+            &mut Self::Empty {
+                ref mut free_list_index,
+            }
+            | &mut Self::Affinity {
+                ref mut free_list_index,
+                ..
+            } => {
+                *free_list_index = index;
+            }
+            _ => panic!("Taken in free list"),
+        }
+    }
+
+    /// Get the index of this slot in its per-module free list.
+    fn per_module_index(&self) -> usize {
+        match self {
+            &Self::Affinity {
+                per_module_index, ..
+            } => per_module_index,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Update the index of this slot in its per-module free list.
+    fn update_per_module_index(&mut self, index: usize) {
+        match self {
+            &mut Self::Affinity {
+                ref mut per_module_index,
+                ..
+            } => {
+                *per_module_index = index;
+            }
+            _ => panic!("Taken in per-module free list"),
```

Comment:
All of the methods on `SlotState` panic if the `SlotState` is `Taken`. What do you think of refactoring `SlotState` to look like this:

```rust
enum SlotState {
    Taken,
    Free(FreeSlotState),
}

impl SlotState {
    // Do the expected matching where `Taken` returns `None` here...
    fn as_free_slot(&self) -> Option<&FreeSlotState> { ... }
    fn as_free_slot_mut(&mut self) -> Option<&mut FreeSlotState> { ... }
}

enum FreeSlotState {
    Empty { ... },
    Affinity { ... },
}

impl FreeSlotState {
    // all the methods above, which don't need to have panicking paths anymore
}
```

This way we can consolidate all the is-this-a-taken-slot checking into a single `slot_state[i].as_free_slot().expect("blah blah blah")` call, our types give us guarantees about their data, and potential panics aren't hidden behind method calls.

Reply:
Could also have `SlotState::unwrap_free_slot[_mut]` methods to avoid the `exepect` at every call site, but still have the potential panics "visible" to callers.

Reply:
Sure, I refactored along these lines (and took the route of panic'ing in the as_free accessors). Good idea, thanks!